src/linker/Linker.Dataflow/XmlFlowAnnotationSource.cs

```diff
@@ -0,0 +1,690 @@
+
+		static readonly string _signature = "signature";
+		static readonly string _fullname = "fullname";
+		static readonly string _preserve = "preserve";
+		static readonly string _accessors = "accessors";
+		static readonly string _ns = string.Empty;
+
+		static readonly string[] _accessorsAll = new string[] { "all" };
+		static readonly char[] _accessorsSep = new char[] { ';' };
+
+		readonly XPathDocument _document;
+		readonly string _xmlDocumentLocation;
+		readonly LinkContext _context;
+
+		public class XmlResolutionException : Exception
+		{
+			public XmlResolutionException (string message, Exception innerException)
+				: base (message, innerException)
+			{
+			}
+		}
+
+		public XmlFlowAnnotationSource (LinkContext context, string document)
+		{
+			_xmlDocumentLocation = document;
+			_document = new XPathDocument (_xmlDocumentLocation);
+			_context = context;
+			Initialize ();
+		}
+
+		public DynamicallyAccessedMemberTypes GetFieldAnnotation (FieldDefinition field)
+		{
+			return _fields.TryGetValue (field, out var ann) ? ann : DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetParameterAnnotation (MethodDefinition method, int index)
+		{
+			if (_methods.TryGetValue (method, out var ann) && ann.ParameterAnnotations != null) {
+				string paramName = method.Parameters[index].Name;
+
+				foreach (var (ParamName, Annotation) in ann.ParameterAnnotations)
+					if (ParamName == paramName)
+						return Annotation;
+			}
+
+			return DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetPropertyAnnotation (PropertyDefinition property)
+		{
+			return _properties.TryGetValue (property, out var ann) ? ann : DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetReturnParameterAnnotation (MethodDefinition method)
+		{
+			return _methods.TryGetValue (method, out var ann) ? ann.ReturnAnnotation : DynamicallyAccessedMemberTypes.None;
+		}
+
+		public DynamicallyAccessedMemberTypes GetThisParameterAnnotation (MethodDefinition method)
+		{
+			if (_methods.TryGetValue (method, out var ann) && ann.ParameterAnnotations != null) {
+
+				foreach (var (ParamName, Annotation) in ann.ParameterAnnotations)
+					if (ParamName == "this")
+						return Annotation;
+			}
+
+			return DynamicallyAccessedMemberTypes.None;
+		}
+
+		static DynamicallyAccessedMemberTypes ParseKinds (ArrayBuilder<Attribute> attributes)
+		{
+			foreach (var attribute in attributes.ToArray ()) {
+				if (attribute.attributeName == "System.Runtime.CompilerServices.DynamicallyAccessedMembers" && attribute.arguments.Count == 1) {
+					foreach (var argument in attribute.arguments.ToArray ()) {
+						if (argument == string.Empty)
+							break;
+						return (DynamicallyAccessedMemberTypes) Enum.Parse (typeof (DynamicallyAccessedMemberTypes), argument);
```

Comment:
This will throw if the value is not a valid enum value - probably leading to a rather nasty error experience (not sure but we probably don't recover from this - worth a try). In any case, this should be reported as a recoverable error (or maybe even a warning? not sure - ideally it should match other error behaviors in our XML parsing logic).

Reply:
A new warning code was introduced for this